src/Planner/PlannerExpressionAnalysis.cpp

```diff
@@ -85,14 +85,14 @@ bool canRemoveConstantFromGroupByKey(const ConstantNode & root)
         else if (function_node)
         {
             /// Do not allow removing constants like `hostName()`
-            if (!function_node->getFunctionOrThrow()->isDeterministic())
+            if (function_node->getFunctionOrThrow()->isServerConstant())
```

Comment:
Will it work for `random` functions?

Reply:
Random does not return constant result, so no constant folding.
For `randConst` we constant fold and substitute result to constant and remove it from GROUP BY key.
```
SELECT
    count(),
    randConstant() AS c1
FROM remote('127.0.0.{3,2}', system, one)
GROUP BY c1
SETTINGS allow_experimental_analyzer = 1

Query id: 15fef62f-a79a-4088-a3b2-44bf5259b2b4

   ┌─count()─┬────────c1─┐
1. │       2 │ 563172653 │
   └─────────┴───────────┘

```

The issue is only for special functions that are not constant on the initiator (so they are not substituted to constants) but constant on the replica and removed from the GROUP BY key.